torch/nn/modules/module.py

```diff
@@ -593,79 +617,103 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
                 persistent buffers.
-            strict (bool): Strictly enforce that the keys in :attr:`state_dict`
-                match the keys returned by this module's `:func:`state_dict()`
-                function.
+            prefix (str): the prefix for parameters and buffers used in this
+                module
+            strict (bool): whether to strictly enforce that the keys in
+                :attr:`state_dict` with :attr:`prefix` match the names of
+                parameters and buffers in this module
+            missing_keys (list of str): if ``strict=False``, add missing keys to
+                this list
+            unexpected_keys (list of str): if ``strict=False``, add unexpected
+                keys to this list
         """
-        def submodule_key_mismatch(full_name, is_missing):
-            module = self
-            names = full_name.split(".")
-            for module_name in names[:-1]:
-                if module_name in module._modules:
-                    module = module._modules[module_name]
-                else:
-                    return
-            module._load_state_dict_key_mismatch(full_name, names[-1], is_missing)
-
-        unexpected = []
-        own_state = self.state_dict()
-        for name, param in state_dict.items():
-            if name in own_state:
-                if isinstance(param, Parameter):
+        local_name_params = itertools.chain(self._parameters.items(), self._buffers.items())
+        local_state = {k: v.data for k, v in local_name_params if v is not None}
+
+        for name, param in local_state.items():
+            key = prefix + name
+            if key in state_dict:
+                input_param = state_dict[key]
+                if isinstance(input_param, Parameter):
                     # backwards compatibility for serialized parameters
-                    param = param.data
+                    input_param = input_param.data
                 try:
-                    own_state[name].copy_(param)
+                    param.copy_(input_param)
                 except Exception:
-                    raise RuntimeError('While copying the parameter named {}, '
+                    raise RuntimeError('While copying the parameter named "{}", '
                                        'whose dimensions in the model are {} and '
                                        'whose dimensions in the checkpoint are {}.'
-                                       .format(name, own_state[name].size(), param.size()))
+                                       .format(key, param.size(), input_param.size()))
             elif strict:
-                unexpected.append(name)
+                missing_keys.append(key)
+
+        if strict:
+            for key, input_param in state_dict.items():
+                if key.startswith(prefix):
+                    input_name = key[len(prefix):]
+                    if '.' not in input_name and input_name not in local_state:
```

Comment:
Why is the `'.' not in input_name` condition here? I'd say that having a dot at this point would be highly unexpected and shouldn't cause us to *not* report an error

Reply:
Because the `key` might be a key for a param in a child of `this` :)